src/lib.rs

```diff
@@ -1,14 +1,95 @@
-//! Rust structs representing network-related types (on Layer 2, 3 and 4) in
-//! Linux.
+//! Rust structs representing network protocol headers (on Layer 2, 3 and 4).
 //!
-//! The crate is [no_std](https://docs.rust-embedded.org/book/intro/no-std.html)
-//! and the structures are fully compatible with the ones provided by the Linux
-//! kernel, which makes it a great fit for [eBPF](https://ebpf.io/) programs
-//! written with [Aya](https://aya-rs.dev/).
+//! The crate is [no_std](https://docs.rust-embedded.org/book/intro/no-std.html),
+//! which makes it a great fit for [eBPF](https://ebpf.io/) programs written
+//! with [Aya](https://aya-rs.dev/).
+//!
+//! # Examples
+//!
+//! An example of an [XDP program](https://aya-rs.dev/book/start/) logging
+//! information about addresses and ports for incoming packets:
+//!
+//! ```rust,ignore
+//! #![no_std]
+//! #![no_main]
+//!
+//! use core::mem;
+//!
+//! use aya_bpf::{bindings::xdp_action, macros::xdp, programs::XdpContext};
+//! use aya_log_ebpf::info;
+//!
+//! use core::mem;
+//! use network_types::{
+//!     eth::{EthHdr, EthProto},
+//!     ip::{Ipv4Hdr, Ipv4Proto},
+//!     tcp::TcpHdr,
+//!     udp::UdpHdr,
+//! };
+//!
+//! const ETH_HDR_LEN: usize = mem::size_of::<EthHdr>();
+//! const IPV4_HDR_LEN: usize = mem::size_of::<Ipv4Hdr>();
+//!
+//! #[xdp]
+//! pub fn xdp_firewall(ctx: XdpContext) -> u32 {
+//!     match try_xdp_firewall(ctx) {
+//!         Ok(ret) => ret,
+//!         Err(_) => xdp_action::XDP_PASS,
+//!     }
+//! }
+//!
+//! #[inline(always)]
+//! unsafe fn ptr_at<T>(ctx: &XdpContext, offset: usize) -> Result<*const T, ()> {
+//!     let start = ctx.data();
+//!     let end = ctx.data_end();
+//!     let len = mem::size_of::<T>();
+//!
+//!     if start + offset + len > end {
+//!         return Err(());
+//!     }
+//!
+//!     Ok((start + offset) as *const T)
+//! }
+//!
+//! fn try_xdp_firewall(ctx: XdpContext) -> Result<u32, ()> {
+//!     let ethhdr: *const EthHdr = unsafe { ptr_at(&ctx, 0)? };
+//!     match unsafe { *ethhdr }.protocol()? {
```

Comment:
protocol() is no more

My advice is that you actually compile these examples or I'm sure there will be errors

Reply:
The problem is that to run those examples, I will have to pull aya-bpf as a dev-dependency.

Reply:
I used this PR branch for my (internal) project, here's the diff [1] - can continue to try building it with the new updates if/once they are added.

[1] https://gist.github.com/dmitris/7ddee1494320b4f6ec2614960bbca0b8

Reply:
Well, ok, actually it's not such a big problem, including it as a dev-dep from git seems to work.

Reply:
It is a problem finally. I'm getting:

```
---- src/lib.rs - (line 12) stdout ----
error: language item required, but not found: `eh_personality`
  |
  = note: this can occur when a binary crate with `#![no_std]` is compiled for a target where `eh_personality` is defined in the standard library
  = help: you may be able to compile for a target that doesn't need `eh_personality`, specify a target with `--target` or in `.cargo/config`

error: aborting due to previous error
```

Understandable, doc tests are compiling for my host target. And sadly, there is no possibility to cross-compile doc tests: https://github.com/rust-lang/cargo/issues/7040

Reply:
I tried using this `.cargo/config`:

```
[build]
target = "bpfel-unknown-none"

[unstable]
build-std = ["core"]
```

but there is an another error:

```
error[E0463]: can't find crate for `test`

error: `#[panic_handler]` function required, but not found

For more information about this error, try `rustc --explain E0463`.
```

Reply:
nvm, I "fixed" that by removing `#![no_std] #![no_main]` from the example, I guess there is no better solution